### PR TITLE
Implement new url structure

### PR DESCRIPTION
**Pull request by ajmaddaford**
### What is the context of this PR?
Implements the URL structure as proposed in https://github.com/ONSdigital/eq-survey-runner/blob/v3/doc/architecture/decisions/0005-simplify-urls-and-extend-to-support-repeating-effectively.md. 

In doing this a large amount of code has been removed:
- Anything to do with repeats
- Household composition and household relationships
- Completeness and dependencies
- Answer summaries (they weren't being used)
- Multiple surveys checking (to be replaced by something appropriate at a later date)

This code has been removed to enable the upcoming feature work for [lists](https://github.com/ONSdigital/eq-survey-runner/blob/v3/doc/architecture/decisions/0006-make-named-lists-a-first-class-construct.md) and hub and spoke navigation.

### How to review 
- Does everything work as expected?
- Has enough been removed?
- Should anything be put back?

### CODE REVIEW

File: app/data_model/answer_store.py [hunk 1]
```diff
@@ -17,29 +17,27 @@ class AnswerStore:
...
-            self.answer_map = existing_answers or defaultdict(list)
+            self.answer_map = existing_answers or {}
 
     def __iter__(self):
-        return iter((answer for answers in self.answer_map.values() for answer in answers))
+        return iter(self.answer_map.values())
 
     def __len__(self):
-        return sum(len(answers) for answers in self.answer_map.values())
+        return len(list(self.answer_map.values()))
```
Comment (jonnyshaw89) [inline app/data_model/answer_store.py]: pretty sure you can just do `len(self.answer_map)`
as key:value is a 121 mapping

File: app/data_model/answer_store.py [hunk 2]
```diff
@@ -118,71 +96,43 @@ def escaped(self):
...
 
         :param answer_ids: The answer ids to filter results by
-        :param answer_instance: The answer instance to filter results by
-        :param group_instance_id: The group instance ID to filter results by
-        :param group_instance: The group instance to filter results by
-        :param limit: True | False Limit the number of answers returned
         :return: Return a new AnswerStore object with filtered answers for chaining
         """
-        filtered = []
+        filtered_map = {k: v for k, v in self.answer_map.items() if k in answer_ids}
```
Comment (jonnyshaw89) [inline app/data_model/answer_store.py]: Can we skip a step with this and do the following?
`answers = [v for k, v in self.answer_map.items() if k in answer_ids]`

File: app/questionnaire/path_finder.py [hunk 3]
```diff
@@ -26,20 +24,7 @@ def __init__(self, schema, answer_store, metadata, completed_blocks):
...
-    @staticmethod
-    def _build_blocks_for_group(group, instance_idx):
-        for block in group['blocks']:
-
-            yield {
-                'group_id': group['id'],
-                'group_instance': instance_idx,
-                'block': block,
-            }
+        return next((index for (index, b) in enumerate(blocks) if b['id'] == location.block_id), None)
```
Comment (jonnyshaw89) [inline app/questionnaire/path_finder.py]: this is me just being annoying but can we change `b` to `block`?

Comment (jonnyshaw89) [reply app/questionnaire/path_finder.py]: #HatesAbbreviations

File: app/questionnaire/path_finder.py [hunk 4]
```diff
@@ -58,38 +43,20 @@ def build_path(self):
...
+            if 'skip_conditions' in group:
+                if evaluate_skip_conditions(group['skip_conditions'], self.schema, self.metadata,
+                                            self.answer_store, routing_path=path):
+                    continue
 
-            if not all_blocks_skipped:
-                if group['id'] in first_groups:
-                    this_location = Location(group['id'], first_group_instance_index, first_block_in_group)
+            for block in group['blocks']:
+                blocks.append(block)
```
Comment (jonnyshaw89) [inline app/questionnaire/path_finder.py]: `blocks.extend(group['blocks'])`

File: app/data_model/answer_store.py [hunk 5]
```diff
@@ -224,31 +174,7 @@ def upgrade_to_1_update_date_formats(answer_store, schema):
...
-            block_id=block_id,
-        )
-        # `get_group_instance_id` handles providing a consistent group_instance_id
-        answer['group_instance_id'] = get_group_instance_id(schema, answer_store, location, answer['answer_instance'])
-
-
 # Dictionary specifying upgrade methods. Key should be the version to upgrade to.
 UPGRADE_TRANSFORMS = {
     1: upgrade_to_1_update_date_formats,
-    2: upgrade_to_2_add_group_instance_id,
```
Comment (bitdivision) [inline app/data_model/answer_store.py:253]: I think we should remove upgrades entirely. We talked about rewriting the interface previously, and these upgrades aren't really relevant on v3.

File: app/submitter/convert_payload_0_0_1.py [hunk 6]
```diff
@@ -1,94 +0,0 @@
-from collections import OrderedDict
-from datetime import datetime
-
-
-def convert_answers_to_payload_0_0_1(answer_store, schema, routing_path):
```
Comment (jonnyshaw89) [inline app/submitter/convert_payload_0_0_1.py]: Why have we removed this in this PR?

File: app/templating/summary_context.py [hunk 7]
```diff
@@ -23,20 +23,11 @@ def build_summary_rendering_context(schema, sections, answer_store, metadata, sc
...
 
     for group in itertools.chain.from_iterable(group_lists):
-        if (group['id'] in group_ids_on_path and schema.group_has_questions(group['id'])):
-            no_of_repeats = _number_of_repeats(group, path)
-            repeating_groups = []
-            for instance_idx in range(0, no_of_repeats):
-                summary_group = Group(group, path, answer_store, metadata, schema, instance_idx, schema_context).serialize()
-                repeating_groups.append(summary_group)
-            groups.extend(repeating_groups)
+        # if (group['id'] in group_ids_on_path and schema.group_has_questions(group['id'])):
```
Comment (jonnyshaw89) [inline app/templating/summary_context.py]: remove comments?

File: app/helpers/form_helper.py [hunk 8]
```diff
@@ -143,47 +80,21 @@ def clear_detail_answer_field(data, questions_for_block):
     return form_data
 
 
-def get_mapped_answers(schema, answer_store, block_id, group_instance, group_instance_id):
+def get_mapped_answers(schema, answer_store, block_id):
     """
     Maps the answers in an answer store to a dictionary of key, value answers. Keys include instance
```
Comment (bitdivision) [inline app/helpers/form_helper.py]: There's a reference to instance id in the docstring here.

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
Looking good

**Review by bitdivision — COMMENTED** [review]
I've looked through all the changes I think. 

Can't see any big issues, all looks good. Happy to approve.

File: app/submitter/convert_payload_0_0_1.py (hunk 6 above)
Comment (ajmaddaford) [reply app/submitter/convert_payload_0_0_1.py]: I got a bit carried away 🙄

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by jonnyshaw89** [comment]
Im happy, you just need to please the Travis / Codecov gods

**Review by bitdivision — APPROVED** [review]
👍

🤞 for the tests

**Pull request merged**